In [1]:
import pandas as pd
import numpy as np
from scipy.sparse import csr_matrix
import sklearn
from sklearn.decomposition import TruncatedSVD

books = pd.read_csv('BX-Books.csv', sep=';', error_bad_lines=False, encoding="latin-1")
books.columns = ['ISBN', 'bookTitle', 'bookAuthor', 'yearOfPublication', 'publisher', 'imageUrlS', 'imageUrlM', 'imageUrlL']
users = pd.read_csv('BX-Users.csv', sep=';', error_bad_lines=False, encoding="latin-1")
users.columns = ['userID', 'Location', 'Age']
ratings = pd.read_csv('BX-Book-Ratings.csv', sep=';', error_bad_lines=False, encoding="latin-1")
ratings.columns = ['userID', 'ISBN', 'bookRating']

c:\users\utshab kumar ghosh\appdata\local\programs\python\python36\lib\importlib\_bootstrap.py:219: RuntimeWarning: numpy.ufunc size changed, may indicate binary incompatibility. Expected 192 from C header, got 216 from PyObject
  return f(*args, **kwds)
c:\users\utshab kumar ghosh\appdata\local\programs\python\python36\lib\importlib\_bootstrap.py:219: RuntimeWarning: numpy.ufunc size changed, may indicate binary incompatibility. Expected 192 from C header, got 216 from PyObject
  return f(*args, **kwds)
b'Skipping line 6452: expected 8 fields, saw 9\nSkipping line 43667: expected 8 fields, saw 10\nSkipping line 51751: expected 8 fields, saw 9\n'
b'Skipping line 92038: expected 8 fields, saw 9\nSkipping line 104319: expected 8 fields, saw 9\nSkipping line 121768: expected 8 fields, saw 9\n'
b'Skipping line 144058: expected 8 fields, saw 9\nSkipping line 150789: expected 8 fields, saw 9\nSkipping line 157128: expected 8 fields, saw 9\nSkipping line 180189: expected 8 fields, saw 9\nSkip

In [3]:
ratings.head()

userID        ISBN  bookRating
0  276725  034545104X           0
1  276726  0155061224           5
2  276727  0446520802           0
3  276729  052165615X           3
4  276729  0521795028           6

In [5]:
users.head()

userID                            Location   Age
0       1                  nyc, new york, usa   NaN
1       2           stockton, california, usa  18.0
2       3     moscow, yukon territory, russia   NaN
3       4           porto, v.n.gaia, portugal  17.0
4       5  farnborough, hants, united kingdom   NaN

In [6]:
books.head()

ISBN                                          bookTitle  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

             bookAuthor yearOfPublication                   publisher  \
0    Mark P. O. Morford              2002     Oxford University Press   
1  Richard Bruce Wright              2001       HarperFlamingo Canada   
2          Carlo D'Este              1991             HarperPerennial   
3      Gina Bari Kolata              1999        Farrar Straus Giroux   
4       E. J. W. Barber              1999  W. W. Norton &amp; Company   

                                           imageUrlS  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                           imageUrlM  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                           imageUrlL  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...  
2  http://images.amazon.com/images/P/0060973129.0...  
3  http://images.amazon.com/images/P/0374157065.0...  
4  http://images.amazon.com/images/P/0393045218.0...

To ensure statistical significance, we will be only looking at the <b>popular books</b>.<br>
In order to find out which books are popular, we need to combine book data with rating data.

In [9]:
combine_book_rating = pd.merge(ratings, books, on='ISBN')
columns = ['yearOfPublication', 'publisher', 'bookAuthor', 'imageUrlS', 'imageUrlM', 'imageUrlL']
combine_book_rating = combine_book_rating.drop(columns, axis = 1)
combine_book_rating.head()

userID        ISBN  bookRating             bookTitle
0  276725  034545104X           0  Flesh Tones: A Novel
1    2313  034545104X           5  Flesh Tones: A Novel
2    6543  034545104X           0  Flesh Tones: A Novel
3    8680  034545104X           5  Flesh Tones: A Novel
4   10314  034545104X           9  Flesh Tones: A Novel

In [10]:
# group by book titles and create a new column for total rating count.
combine_book_rating = combine_book_rating.dropna(
axis = 0, subset = ['bookTitle'])

In [13]:
book_ratingCount = (combine_book_rating.
                   groupby(by = ['bookTitle'])['bookRating'].
                   count().reset_index().
                   rename(columns = {'bookRating': 'totalRatingCount'})
                   [['bookTitle', 'totalRatingCount']]
                   )
book_ratingCount.head()

bookTitle  totalRatingCount
0   A Light in the Storm: The Civil War Diary of ...                 4
1                              Always Have Popsicles                 1
2               Apple Magic (The Collector's series)                 1
3   Ask Lily (Young Women of Faith: Lily Series, ...                 1
4   Beyond IBM: Leadership Marketing and Finance ...                 1

In [15]:
# combine the rating data with the total rating count data,
# this gives us exactly what we need to filter out the lesser known books.

rating_with_totalRatingCount = combine_book_rating.merge(
    book_ratingCount, left_on = 'bookTitle', right_on = 'bookTitle',how = 'inner')
rating_with_totalRatingCount.head()

userID        ISBN  bookRating             bookTitle  totalRatingCount
0  276725  034545104X           0  Flesh Tones: A Novel                60
1    2313  034545104X           5  Flesh Tones: A Novel                60
2    6543  034545104X           0  Flesh Tones: A Novel                60
3    8680  034545104X           5  Flesh Tones: A Novel                60
4   10314  034545104X           9  Flesh Tones: A Novel                60

In [16]:
# statistics of total rating count:
pd.set_option('display.float_format', lambda x: '%.3f' %x)
print(book_ratingCount['totalRatingCount'].describe())

count   241071.000
mean         4.277
std         16.739
min          1.000
25%          1.000
50%          1.000
75%          3.000
max       2502.000
Name: totalRatingCount, dtype: float64


In [17]:
# The median book has been rated only once. 
# Let’s look at the top of the distribution:
print(book_ratingCount['totalRatingCount'].quantile
      (np.arange(.9, 1, .01)))

0.900    7.000
0.910    8.000
0.920    9.000
0.930   10.000
0.940   11.000
0.950   13.000
0.960   16.000
0.970   20.000
0.980   29.000
0.990   50.000
Name: totalRatingCount, dtype: float64


About 1% of the books received 50 or more ratings. Because we have so many books in our data, we will limit it to the top <b>1%</b>, and this will give us <b>2713 unique books</b>.

In [20]:
popularity_threshold= 50
rating_popular_book = rating_with_totalRatingCount.query(
    'totalRatingCount >= @popularity_threshold')
rating_popular_book.head()

userID        ISBN  bookRating             bookTitle  totalRatingCount
0  276725  034545104X           0  Flesh Tones: A Novel                60
1    2313  034545104X           5  Flesh Tones: A Novel                60
2    6543  034545104X           0  Flesh Tones: A Novel                60
3    8680  034545104X           5  Flesh Tones: A Novel                60
4   10314  034545104X           9  Flesh Tones: A Novel                60

In [24]:
# Filter to users in US and Canada only
combined = rating_popular_book.merge(users, 
                                     left_on='userID',
                                     right_on='userID',
                                     how = 'left')

us_canada_user_rating = combined[
    combined['Location'].str.contains('usa|canada')]
us_canada_user_rating = us_canada_user_rating.drop('Age', axis = 1)
us_canada_user_rating.head()

userID        ISBN  bookRating             bookTitle  totalRatingCount  \
0  276725  034545104X           0  Flesh Tones: A Novel                60   
1    2313  034545104X           5  Flesh Tones: A Novel                60   
2    6543  034545104X           0  Flesh Tones: A Novel                60   
3    8680  034545104X           5  Flesh Tones: A Novel                60   
4   10314  034545104X           9  Flesh Tones: A Novel                60   

                            Location  
0                  tyler, texas, usa  
1              cincinnati, ohio, usa  
2           strafford, missouri, usa  
3  st. charles county, missouri, usa  
4             beaverton, oregon, usa

<h3>Implementing kNN</h3>

We convert our table to a 2D matrix, and fill the missing values with zeros (since we will calculate distances between rating vectors). We then transform the values(ratings) of the matrix dataframe into a scipy sparse matrix for more efficient calculations.

In [33]:
#Finding the Nearest Neighbors
us_canada_user_rating = us_canada_user_rating.drop_duplicates(['userID', 'bookTitle'])
us_canada_user_rating_pivot = us_canada_user_rating.pivot(index='bookTitle',columns = 'userID',
                                                          values = 'bookRating').fillna(0)
us_canada_user_rating_matrix = csr_matrix(us_canada_user_rating_pivot.values)

In [34]:
from sklearn.neighbors import NearestNeighbors

model_knn = NearestNeighbors(metric = 'cosine', algorithm='brute')
model_knn.fit(us_canada_user_rating_matrix)

NearestNeighbors(algorithm='brute', leaf_size=30, metric='cosine',
                 metric_params=None, n_jobs=None, n_neighbors=5, p=2,
                 radius=1.0)

In [43]:
# Test our model and make some recommendations:
query_index = np.random.choice(us_canada_user_rating_pivot.shape[0])
distances, indices = model_knn.kneighbors(us_canada_user_rating_pivot.iloc[query_index, :].values.reshape(1, -1), n_neighbors = 6)

for i in range(0, len(distances.flatten())):
    if i==0:
        print("Recommendations for {0}:\n".format(
            us_canada_user_rating_pivot.index[query_index]))
    else:
        print('{0}: {1}, with distance of {2}'.format(
            i, us_canada_user_rating_pivot.index[indices.flatten()[i]], distances.flatten()[i]))
        

Recommendations for The Genesis Code:

1: Trevayne, with distance of 0.8386032789693919
2: 2061: Odyssey Three, with distance of 0.8737323769409746
3: Undue Influence, with distance of 0.875343168685911
4: The Worst-Case Scenario Survival Handbook, with distance of 0.8773234531605988
5: Word Freak: Heartbreak, Triumph, Genius, and Obsession in the World of Competitive Scrabble Players, with distance of 0.9010193717194313


<h3>Collaborative Filtering Using Matrix Factorization</h3>
The Matrix Factorization techniques are usually more effective, because they allow users to discover the latent (hidden)features underlying the interactions between users and items (books).<br><br>
We use singular value decomposition (SVD) — one of the Matrix Factorization models for identifying latent factors.<br><br>
Similar with kNN, we convert our USA Canada user rating table into a 2D matrix (called a utility matrix here) and fill the missing values with zeros.

In [45]:
us_canada_user_rating_pivot2 = us_canada_user_rating.pivot(index = 'userID',
                                                          columns = 'bookTitle',
                                                          values = 'bookRating').fillna(0)
us_canada_user_rating_pivot2.head()

bookTitle  10 Lb. Penalty  16 Lighthouse Road  1984  1st to Die: A Novel  \
userID                                                                     
8                   0.000               0.000 0.000                0.000   
9                   0.000               0.000 0.000                0.000   
14                  0.000               0.000 0.000                0.000   
16                  0.000               0.000 0.000                0.000   
17                  0.000               0.000 0.000                0.000   

bookTitle  2010: Odyssey Two  204 Rosewood Lane  2061: Odyssey Three  \
userID                                                                 
8                      0.000              0.000                0.000   
9                      0.000              0.000                0.000   
14                     0.000              0.000                0.000   
16                     0.000              0.000                0.000   
17                     0.000              0.000                0.000   

bookTitle  24 Hours  2nd Chance  3rd Degree  ...  YOU BELONG TO ME  \
userID                                       ...                     
8             0.000       0.000       0.000  ...             0.000   
9             0.000       0.000       0.000  ...             0.000   
14            0.000       0.000       0.000  ...             0.000   
16            0.000       0.000       0.000  ...             0.000   
17            0.000       0.000       0.000  ...             0.000   

bookTitle  Year of Wonders  You Belong To Me  You Shall Know Our Velocity  \
userID                                                                      
8                    0.000             0.000                        0.000   
9                    0.000             0.000                        0.000   
14                   0.000             0.000                        0.000   
16                   0.000             0.000                        0.000   
17                   0.000             0.000                        0.000   

bookTitle  Young Wives  \
userID                   
8                0.000   
9                0.000   
14               0.000   
16               0.000   
17               0.000   

bookTitle  Zen and the Art of Motorcycle Maintenance: An Inquiry into Values  \
userID                                                                         
8                                                      0.000                   
9                                                      0.000                   
14                                                     0.000                   
16                                                     0.000                   
17                                                     0.000                   

bookTitle  Zoya  \O\" Is for Outlaw"  \
userID                                 
8         0.000                0.000   
9         0.000                0.000   
14        0.000                0.000   
16        0.000                0.000   
17        0.000                0.000   

bookTitle  \Surely You're Joking, Mr. Feynman!\": Adventures of a Curious Character"  \
userID                                                                                 
8                                                      0.000                           
9                                                      0.000                           
14                                                     0.000                           
16                                                     0.000                           
17                                                     0.000                           

bookTitle  stardust  
userID               
8             0.000  
9             0.000  
14            0.000  
16            0.000  
17            0.000  

[5 rows x 2442 columns]

We then transpose this utility matrix, so that the bookTitles become rows and userIDs become columns. After using TruncatedSVD to decompose it, we fit it into the model for dimensionality reduction. This compression happened on the dataframe’s columns since we must preserve the book titles. We choose n_components = 12 for just 12 latent variables, and you can see, our data’s dimensions have been reduced significantly from 40017 X 2442 to 2442 X 12.

In [46]:
us_canada_user_rating_pivot2.shape

(40017, 2442)

In [47]:
X = us_canada_user_rating_pivot2.values.T
X.shape

(2442, 40017)

In [48]:
from sklearn.decomposition import TruncatedSVD

SVD = TruncatedSVD(n_components=12, random_state=17)
matrix = SVD.fit_transform(X)
matrix.shape

(2442, 12)

We calculate the <b>Pearson’s R correlation coefficient</b> for every book pair in our final matrix. To compare this with the results from kNN, we pick the same book <b>“The Genesis Code”</b> to find the books that have high correlation coefficients (between 0.9 and 1.0) with it.

In [49]:
import warnings
warnings.filterwarnings('ignore', category = RuntimeWarning)
corr = np.corrcoef(matrix)
corr.shape

(2442, 2442)

In [51]:
us_canada_book_title = us_canada_user_rating_pivot2.columns
us_canada_book_list = list(us_canada_book_title)
coffey_hands = us_canada_book_list.index("The Genesis Code")
print(coffey_hands)

1885


In [52]:
corr_coffey_hands = corr[coffey_hands]

In [53]:
list(us_canada_book_title[(corr_coffey_hands<1.0) & (corr_coffey_hands>0.9)])

['Devices and Desires',
 'Finding Moon',
 'Ground Zero and Beyond',
 'The Charm School',
 'The Wailing Wind',
 'Up Country']

<h1>References</h1>
1. <a>https://towardsdatascience.com/how-did-we-build-book-recommender-systems-in-an-hour-part-2-k-nearest-neighbors-and-matrix-c04b3c2ef55c</a><br><br>
2. <a>https://github.com/RezaGhanbari/book-sharing-recommender/blob/master/main.py</a>